go-controller/pkg/node/node.go

```diff
@@ -286,54 +292,45 @@ func (n *OvnNode) WatchEndpoints() {
 		UpdateFunc: func(old, new interface{}) {
 			epNew := new.(*kapi.Endpoints)
 			epOld := old.(*kapi.Endpoints)
+			// return if there are no endpoint changes
 			if apiequality.Semantic.DeepEqual(epNew.Subsets, epOld.Subsets) {
 				return
 			}
-			newEpAddressMap := buildEndpointAddressMap(epNew.Subsets)
-			for item := range buildEndpointAddressMap(epOld.Subsets) {
-				if _, ok := newEpAddressMap[item]; !ok {
-					err := deleteConntrack(item.ip, item.port, item.protocol)
-					if err != nil {
-						klog.Errorf("Failed to delete conntrack entry for %s: %v", item.ip, err)
-					}
-				}
+			// get the service that belongs to the endpoint
+			svc, err := n.watchFactory.GetService(epNew.Namespace, epNew.Name)
```

Comment:
Haha, doh! I gave you a comment on your other PR. But anyways, I will put it here instead: https://github.com/ovn-org/ovn-kubernetes/pull/1857/commits/7ba22026569334d65648c7517c61b59220838ba8#r531525158

> Can't we change this to a service watcher instead? Endpoints are not needed anymore it seems.

Reply:
we need to detect events when endpoints change , a service A can have an endpointB, if this is deleted and a new endpointC is created, we never notice watching only services. Also, we need to delete the stale entries to endpointB so endpointC receive the traffic, that is exactly what the e2e that Jacob added in upstream does.

Reply:
Oh! But in that case when we start supporting endpoint slices, we'll need the permissions on ovnkube-node's side (according to what we discussed here: https://github.com/openshift/cluster-network-operator/pull/891#discussion_r531102953) to do the same, right?

Reply:
not really, endpoints will be always present, we need to detect that there is a change in endpoints, doesn't matter watching slices or endpoints objects

Reply:
OK, so you're saying that with the new endpoint API whenever a endpoint slice object is updated it also updates the endpoint object? I understood that they will be de-correlated. This is what confuses me here: if we only watch endpoints how can we be sure that we do what is necessary for the endpoint slice?

Reply:
let me start over, the endpoints and endpoint slices are generated by controllers that run in the `kube-controller-manager`, both are independent. This controllers watch the services with selector and generate `endpoint` or `endpointSlices` objects with the pods Addresses that matches those selector.

The endpointSlice controller is the only one that generate endpoints for each IP family of the pod address, endpoints only pick one address of each pod. If some pod address needs to be added or removed from one endpoint, it has to happen the same with the slice, and viceversa.

Reply:
OK, that is a bit clearer. So essentially one affects the other, and vice versa - so it's enough to watch one of them.  From [1] I have seen the following noted:

```
The control plane mirrors Endpoints resources unless:

the Endpoints resource has a endpointslice.kubernetes.io/skip-mirror label set to true.
the Endpoints resource has a control-plane.alpha.kubernetes.io/leader annotation.
the corresponding Service resource does not exist.
the corresponding Service resource has a non-nil selector.
```

I guess we don't care about the bottom two cases. But how about the first two? The mirroring would be broken for them and we might miss endpoint slice conntrack entries, do we care about that?  

[1]: https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#endpointslice-mirroring

Reply:
hehe, I tried to avoid this but you've found it, there is another controller, the endpoints slice mirrororing controller, this is needed to have compatibility with the applications that create endpoints ONLY, but we don't care about those, we only need the endpoints that are autogenerated due to the pod addresses